patina_dxe_core/src/allocator.rs

```diff
@@ -328,10 +328,42 @@ impl AllocatorMap {
                 NonNull::from_ref(Box::leak(Box::new(EFiMemoryTypeInformation { memory_type, number_of_pages: 0 })))
             };
 
-            Box::leak(Box::new(UefiAllocator::new(&GCD, memory_type_info, handle, granularity)))
+            let allocator = Box::leak(Box::new(UefiAllocator::new(&GCD, memory_type_info, handle, granularity)));
+
+            // Initialize the allocator with existing memory regions of this type from the GCD.
+            Self::initialize_allocator_with_existing_regions(allocator, handle);
+
+            allocator
         })
     }
 
+    /// Initializes a newly-created allocator with existing memory regions from the GCD
+    /// that match the allocator's handle.
+    ///
+    /// This prevents the MIN_EXPANSION policy from being triggered on first allocation
+    /// when suitable memory already exists.
+    fn initialize_allocator_with_existing_regions(allocator: &UefiAllocator, handle: efi::Handle) {
+        let mut descriptors: Vec<dxe_services::MemorySpaceDescriptor> =
+            Vec::with_capacity(GCD.memory_descriptor_count() + 10);
+        if GCD.get_memory_descriptors(&mut descriptors).is_err() {
+            // Just proceed without initializing if there is a problem retrieving descriptors
+            return;
+        }
+
+        for descriptor in descriptors.iter() {
```

Comment:
I think you need to differentiate between free and allocated descriptors in the GCD. 

Might be easier to just call GCD.allocate_memory_space(). I believe that the GCD logic will already attempt to satisfy the allocation with pages that are assigned to this memory type if it can.